cloud/scope/machine_pool.go

```diff
@@ -0,0 +1,510 @@
+
+	ads := s.OCICluster.Status.AvailabilityDomains
+
+	specPlacementDetails := s.OCIMachinePool.Spec.PlacementDetails
+
+	// make sure user doesn't specify 3 ads when there is only one available
+	if len(specPlacementDetails) > len(ads) {
+		errMsg := fmt.Sprintf("Cluster has %d ADs specified and the machine pools spec has %d", len(ads), len(specPlacementDetails))
+		return nil, errors.New(errMsg)
+	}
+
+	// build placements from the user spec
+	for _, ad := range ads {
+		for _, specPlacment := range specPlacementDetails {
+			if strings.HasSuffix(ad.Name, strconv.Itoa(specPlacment.AvailabilityDomain)) {
+				placement := core.CreateInstancePoolPlacementConfigurationDetails{
+					AvailabilityDomain: common.String(ad.Name),
+					PrimarySubnetId:    s.GetWorkerMachineSubnet(),
+					FaultDomains:       ad.FaultDomains,
+				}
+				s.Info("Adding machine placement for AD", "AD", ad.Name)
+				placements = append(placements, placement)
+			}
+		}
+	}
+
+	// build placements if the user hasn't specified any
+	if len(placements) <= 0 {
+		for _, ad := range ads {
+			placement := core.CreateInstancePoolPlacementConfigurationDetails{
+				AvailabilityDomain: common.String(ad.Name),
+				PrimarySubnetId:    s.GetWorkerMachineSubnet(),
+				FaultDomains:       ad.FaultDomains,
+			}
+			placements = append(placements, placement)
+		}
+	}
+
+	return placements, nil
+}
+
+// IsResourceCreatedByClusterAPI determines if the instance was created by the cluster using the
+// tags created at instance launch.
+func (s *MachinePoolScope) IsResourceCreatedByClusterAPI(resourceFreeFormTags map[string]string) bool {
+	tagsAddedByClusterAPI := ociutil.BuildClusterTags(string(s.OCICluster.GetOCIResourceIdentifier()))
+	for k, v := range tagsAddedByClusterAPI {
+		if resourceFreeFormTags[k] != v {
+			return false
+		}
+	}
+	return true
+}
+
+// GetFreeFormTags gets the free form tags for the MachinePoolScope cluster and returns them
+func (m *MachinePoolScope) GetFreeFormTags() map[string]string {
+
+	tags := ociutil.BuildClusterTags(m.OCICluster.GetOCIResourceIdentifier())
+	if m.OCICluster.Spec.FreeformTags != nil {
+		for k, v := range m.OCICluster.Spec.FreeformTags {
+			tags[k] = v
+		}
+	}
+
+	return tags
+}
+
+// ReconcileInstanceConfiguration works to try to reconcile the state of the instance configuration for the cluster
+func (m *MachinePoolScope) ReconcileInstanceConfiguration(ctx context.Context) error {
+	var instanceConfiguration *core.InstanceConfiguration
+	instanceConfigurationId := m.GetInstanceConfigurationId()
+
+	if len(instanceConfigurationId) > 0 {
+		req := core.GetInstanceConfigurationRequest{InstanceConfigurationId: common.String(instanceConfigurationId)}
+		instanceConfiguration, err := m.ComputeManagementClient.GetInstanceConfiguration(ctx, req)
+		if err == nil {
+			m.Info("instance configuration found", "InstanceConfigurationId", instanceConfiguration.Id)
+			m.SetInstanceConfigurationIdStatus(instanceConfigurationId)
```

Comment:
what happens if the ocid for the image changes, for example via kubernetes upgrade?

Reply:
Will be addressed in https://github.com/oracle/cluster-api-provider-oci/issues/90